components/provisioner/internal/parallel-installation/components_status.go

```diff
@@ -0,0 +1,83 @@
+package parallel_installation
+
+import (
+	"fmt"
+
+	"github.com/kyma-project/control-plane/components/provisioner/internal/model"
+
+	"github.com/kyma-incubator/hydroform/parallel-install/pkg/components"
+	"github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment"
+)
+
+type ComponentsStatus struct {
+	components         map[string]bool
+	componentFailed    error
+	installationFailed error
+}
+
+func NewComponentsStatus(components []model.KymaComponentConfig) *ComponentsStatus {
+	cmpList := make(map[string]bool)
+	for _, component := range components {
+		cmpKey := componentKey(string(component.Component), component.Namespace)
+		cmpList[cmpKey] = false
+	}
+
+	return &ComponentsStatus{components: cmpList}
+}
+
+func (cs *ComponentsStatus) ConsumeEvent(event deployment.ProcessUpdate) {
+	// check if installation process failed
+	if event.Event == deployment.ProcessExecutionFailure ||
```

Comment:
During tests I came across the following crash:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation cod
e=0x1 addr=0x0 pc=0x1d60bd2]

goroutine 6737 [running]:
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.(*ComponentsStatus).ConsumeEvent(0x0, 0x239565b, 0xe, 0x239ac17, 0x11, 0x0, 0x0, 0xc004a17110, 0x15, 0xc000b9ba00, ...)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/components_status.go:38 +0x152
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.parallelInstallationService.getCallbackUpdate.func2(0x239565b, 0xe, 0x239ac17, 0x11, 0x0, 0x0, 0xc004a17110, 0x15, 0xc000b9ba00, 0x10, ...)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/parallel_installation.go:84 +0xbc
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.parallelInstallationService.getCallbackUpdate.func3(0x239565b, 0xe, 0x239ac17, 0x11, 0x0, 0x0, 0xc004a17110, 0x15, 0xc000b9ba00, 0x10, ...)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/parallel_installation.go:95 +0x335
github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment.(*core).processUpdateComponent(0xc002690810, 0x239ac17, 0x11, 0xc004a17110, 0x15, 0xc000b9ba00, 0x10, 0xc000b9b100, 0xa, 0x238ceea, ...)
	/go/pkg/mod/github.com/kyma-incubator/hydroform/parallel-install@v0.0.0-20210610074205-bbaf6958f1a5/pkg/deployment/core.go:115 +0x155


github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment.(*Deployment).deployComponents(0xc0022681a8, 0x26af3c0, 0xc0071455c0, 0xc007354840, 0x239ac17, 0x11, 0xc007145580, 0x1174f2ca8ea, 0x15d289160ea, 0x0, ...)
	/go/pkg/mod/github.com/kyma-incubator/hydroform/parallel-install@v0.0.0-20210610074205-bbaf6958f1a5/pkg/deployment/deployment.go:153 +0x358
github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment.(*Deployment).startKymaDeployment(0xc0022681a8, 0x2677b78, 0xc007145500, 0xc007145540, 0xc007145580, 0x0, 0x0)
	/go/pkg/mod/github.com/kyma-incubator/hydroform/parallel-install@v0.0.0-20210610074205-bbaf6958f1a5/pkg/deployment/deployment.go:130 +0x5d8
github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment.(*Deployment).StartKymaDeployment(0xc0022681a8, 0xc006652240, 0xc001987a90)
	/go/pkg/mod/github.com/kyma-incubator/hydroform/parallel-install@v0.0.0-20210610074205-bbaf6958f1a5/pkg/deployment/deployment.go:85 +0x2d6
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.(*Deployer).StartKymaDeployment(0xc00263a1e0, 0x0, 0x0)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/deployer.go:24 +0x7a
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.AsyncDeployment.StartKymaDeployment.func1(0x2661c60, 0xc00263a1e0, 0x2494ff8, 0x2495000)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/parallel_installation.go:40 +0x31
created by github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.AsyncDeployment.StartKymaDeployment
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/parallel_installation.go:39 +0x5d
```

I suppose that this behaviour is caused by the fact, that CheckInstallationStatus removes entry for RuntimeID when the installation fails or succeeds. If there is an event sent by parallel-install after installation completed event object will be nil. We can add nil checks to solve the issue, but parallel-install shouldn't sent any update after installation finished.

Reply:
I updated `consumeEvent` function inside the callback method:
```go
consumeEvent := func(event deployment.ProcessUpdate) {
   p.mux.Lock()
   if _, ok := p.installationStatus[runtimeID]; ok {
      p.installationStatus[runtimeID].ConsumeEvent(event)
   } else {
      p.log.Warnf("status checker for runtime: %s not exist anymore", runtimeID)
   }
   p.mux.Unlock()
}
```
when the installation status checker with specific runtimeID will be removed then `ConsumeEvent` method won't be triggered.
I also add immidietally return from `ConsumeEvent` if event type suggests installation error.